### ISSUE TITLE
flakes: git filters broken

### ISSUE THREAD

**Issue by alarsyo** [NixOS/nix#5260]
**Describe the bug**

If you have a problem with a specific package or NixOS,
you probably want to file an issue at https://github.com/NixOS/nixpkgs/issues. 

**Steps To Reproduce**

**Expected behavior**

**`nix-env --version` output**

**Additional context**

**Comment by nrdxp** [comment]
why would fetching submodules break git-crypt? Do you have your secrets in a submodule?

**Comment by plabadens** [comment]
I'm able to reproduce the issue. Somehow this has changed the way the git repo gets moved to the nix store.

**Comment by nrdxp** [comment]
oh I guess I can see that, since it probably copies the repo from the git history (which has the secrets encrypted by default), instead of just a plain copy. For the sake of investigation, is this broken even when your file tree is dirty? I would imagine the behavior would be different in this case since it can't simply rely on the commit history since dirty changes don't exist there yet.

**Comment by plabadens** [comment]
I'm currently doing a bit of testing, but it seems to evaluate perfectly fine when the worktree is dirty, as you expected. Using `git-crypt` was always a bit of a hack and I should probably be moving to something like `sops-nix` anyway.

**Comment by alarsyo** [comment]
Yes, I should also move to sops-nix or age-nix, but having a dirty worktree is indeed a great workaround for the time being, thanks for the tip!

**Comment by miuirussia** [comment]
I also encountered this bug, rolled back the changes that led to an error using the overlay: https://github.com/miuirussia/nix-flake-env/blob/master/overlays/nixUnstable/default.nix

**Comment by nuance** [comment]
I was using git-lfs to manage some binaries in my nix config, which this also appears to break. Reverting the commit fixes the issue.

**Comment by nrdxp** [comment]
the commit is reverted in master so we can close this I think.

**Comment by roberth** [comment]
Git-crypt should _not_ work, at least not with the way flakes are currently implemented. Nix Flakes will write your secrets unconditionally to the store, which is not a secure place.
Also, like arbitrary smudge filters, it is not reproducible. See https://github.com/NixOS/nix/pull/4635 for a more detailed argumentation.

I would like for git-crypt to work _somehow_, but it would take special treatment of local flakes, which I think it is too early to consider that at this point.

**Comment by lunik1** [comment]
Something of a workaround for the `git-crypt` case is to ensure the encrypted file is dirty before you evaluate the flake; this way the unencrypted file is evaluated. Of course, your secrets will still be globally readable in the nix store.

**Comment by roberth** [comment]
A proper git-crypt integration both allows access to the encrypted file for delayed decryption, and integrates with the string context to avoid adding decrypted files to the store
 - #8388

**Comment by chayleaf** [comment]
Disagreed, adding decrypted files to store is fine for some use cases - i.e. where you're fine with the files being in your locally readable store, but not in your online git repo.

**Comment by roberth** [comment]
@chayleaf would you be ok with a function that turns a secret string into a regular string?
While this does remove an absolute guarantee from the system, it does still prevent almost all _accidental_ leaking, assuming that libraries will not apply this function for you, which is reasonable.
I would not reuse `builtins.unsafeDiscardStringContext` for this, because code has been written without this consideration in mind. It should be explicit about secrets; for example: `builtins.exposeSecret`.

**Comment by chayleaf** [comment]
I think the best option is to have the secret *files* marked as encrypted, but have any *processing* of the file (e.g. import, fromJSON) lose that context, it would be hard to do otherwise anyway
